### PR TITLE
feat(cli): node llama cpp

### LINKED ISSUE #52: Does not work on ubuntu 22.04 & node 20
Please refer to the [troubleshooting](https://github.com/ido-pluto/catai/blob/main/docs/troubleshooting.md) before opening an issue. You might find the solution there.

**Describe the bug**
Followed instructions, when `catai up` is used it only opens a page for 1 time, then process exits.

**Screenshots**
```
$ catai up
CatAI client on http://127.0.0.1:3000
New connection
$ echo $?
0
```

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - Browser [e.g. chrome, safari] Chrome
 - CatAI version [e.g. 0.3.10] (`catai --version`) 3.0.0
 - Node.js version [e.g 19] (`node --version`) 20

### PR DESCRIPTION
**Pull request by ido-pluto**
### Description of change
Add node-llama-cpp CLI to catai
- Option to recompile binaries in case they broken
- Troubleshooting common problems

Fixes #52

### Pull-Request Checklist

- [x] Code is up-to-date with the `main` branch
- [ ] This pull request links relevant issues as `Fixes #0000`
- [x] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained
  in [CONTRIBUTING.md](https://github.com/withcatai/catai/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: server/src/cli/commands/node-llama-cpp.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import {createRequire} from "module";
+import fs from 'fs-extra';
+import path from 'path';
+
+const NODE_LLAMA_CPP_LOCATION = createRequire(import.meta.url).resolve('node-llama-cpp');
+const BINARIES_GITHUB_RELEASE_JSON = path.resolve(`${NODE_LLAMA_CPP_LOCATION}/../../llama/binariesGithubRelease.json`);
+const {release: DEFAULT_RELEASE} = await fs.readJson(BINARIES_GITHUB_RELEASE_JSON);
+
+const METAL_ENABLED_BY_DEFAULT = process.platform === 'darwin';
+const LLAMA_CPP_REPO = 'ggerganov/llama.cpp';
```
Comment (giladgd) [inline server/src/cli/commands/node-llama-cpp.ts]: This is a bad idea. Do not depend on the internals of `node-llama-cpp`.
There's no need to have a copy of the defaults here, just pass `undefined` to use the defaults managed by `node-llama-cpp`

File: server/src/cli/commands/node-llama-cpp.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+nodeLlamaCpp.description('Node llama.cpp CLI - recompile node-llama-cpp binaries')
+    .alias('cpp')
+    .option('--repo <repo>', 'The GitHub repository to download a release of llama.cpp from')
+    .option('--release <release>', 'The tag of the llama.cpp release to download. Set to "latest" to download the latest release')
+    .option('--arch <arch>', 'The architecture to compile llama.cpp for')
+    .option('--nodeTarget <nodeTarget>', 'The Node.js version to compile llama.cpp for. Example: v18.0.0')
+    .option('--metal', 'Compile llama.cpp with Metal support. Enabled by default on macOS')
+    .option('--cuda', 'Compile llama.cpp with CUDA support')
+    .option('--no-download', 'Skip downloading llama.cpp and only build it');
+nodeLlamaCpp.action(async ({repo = LLAMA_CPP_REPO, release = DEFAULT_RELEASE, arch, nodeTarget, cuda, metal = METAL_ENABLED_BY_DEFAULT, noDownload = false}) => {
```
Comment (giladgd) [inline server/src/cli/commands/node-llama-cpp.ts]:
```suggestion
nodeLlamaCpp.action(async ({repo, release, arch, nodeTarget, cuda, metal, noDownload = false}) => {
```

**Review by giladgd — CHANGES_REQUESTED** [review]

File: server/src/cli/commands/node-llama-cpp.ts (hunk 1 above)
Comment (ido-pluto) [reply server/src/cli/commands/node-llama-cpp.ts]: That will not work, I checked that:
image

Comment (giladgd) [reply server/src/cli/commands/node-llama-cpp.ts]: You're right; my bad.
I've fixed it in this PR: https://github.com/withcatai/node-llama-cpp/pull/91

**Pull request merged**